pkg/ca/intermediateca/intermediateca.go

```diff
@@ -0,0 +1,215 @@
+							},
+							ExtendedKeyUsage: &privatecapb.KeyUsage_ExtendedKeyUsageOptions{
+								CodeSigning: true,
+							},
+						},
+						CaOptions: &privatecapb.X509Parameters_CaOptions{
+							IsCa:                &isCa,
+							MaxIssuerPathLength: &maxIssuerPathLength,
+						},
+					},
+					SubjectConfig: &privatecapb.CertificateConfig_SubjectConfig{
+						Subject: &privatecapb.Subject{
+							CommonName:   "sigstore-intermediate",
+							Organization: "sigstore.dev",
+						},
+					},
+				},
+			},
+		},
+	}
+
+	resp, err := client.CreateCertificate(ctx, csr)
+	if err != nil {
+		return nil, err
+	}
+
+	var pemCerts []string
+	pemCerts = append(pemCerts, resp.PemCertificate)
+	pemCerts = append(pemCerts, resp.PemCertificateChain...)
+
+	var parsedCerts []*x509.Certificate
+	for _, c := range pemCerts {
+		certs, err := cryptoutils.UnmarshalCertificatesFromPEM([]byte(c))
+		if err != nil {
+			return nil, err
+		}
+		if len(certs) != 1 {
+			return nil, errors.New("unexpected number of certificates returned")
+		}
+		parsedCerts = append(parsedCerts, certs[0])
+	}
+	log.Logger.Info("Current CA certificate chain:")
+	log.Logger.Info(pemCerts)
+
+	return parsedCerts, nil
+}
+
+func refreshCACertificate(ctx context.Context, ica *intermediateCA) {
+	ticker := time.NewTicker(time.Hour * 4)
+	for range ticker.C {
+		ica.RLock()
+		currentCert := ica.certs[0]
+		ica.RUnlock()
+
+		// Refresh certificate 7 days before expiration
+		if time.Until(currentCert.NotAfter) < (time.Hour * 24 * 7) {
+			certs, err := fetchCACertificate(ctx, ica.parent, ica.client, ica.signer)
+			if err != nil {
+				// An intermittent error is acceptable
+				log.Logger.Error(err)
+				continue
+			}
+			ica.updatedCerts <- certs
+		}
+	}
+}
+
+func updateCACertificate(ica *intermediateCA) {
+	for certs := range ica.updatedCerts {
+		ica.Lock()
+		ica.certs = certs
+		ica.Unlock()
+	}
+}
+
+func NewIntermediateCA(ctx context.Context, parent string) (ca.CertificateAuthority, error) {
+	var ica intermediateCA
+
+	signer, _, err := signature.NewDefaultECDSASignerVerifier()
+	if err != nil {
+		return nil, err
+	}
+	ica.signer = signer
+
+	client, err := privateca.NewCertificateAuthorityClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+	ica.client = client
+	ica.parent = parent
+
+	ica.certs, err = fetchCACertificate(ctx, ica.parent, ica.client, ica.signer)
+	if err != nil {
+		return nil, err
+	}
+
+	ica.updatedCerts = make(chan []*x509.Certificate)
+
+	// Start goroutine to periodically check and refresh CA certificate and chain
+	go refreshCACertificate(ctx, &ica)
+	// Start goroutine to update CA certificate and chain
+	go updateCACertificate(&ica)
+
+	return &ica, nil
+}
+
+func (ica *intermediateCA) CreateCertificate(ctx context.Context, challenge *challenges.ChallengeResult) (*ca.CodeSigningCertificate, error) {
+	cert, err := x509ca.MakeX509(challenge)
+	if err != nil {
+		return nil, err
+	}
+
+	parentCA, privateKey := ica.getX509KeyPair()
```

Comment:
since the locking is done within this function call, wouldn't it be possible that after returning from line 192, `ica.certs` would be updated in the background before getting to line 199? 

that would end up in a situation where the value of `ica.certs` changes from what was used to create the CSC vs the chain reported back in the response message to the caller. 

why not just change line 214 to return all `ica.certs`, and then use that local copy instead of going back to a read lock and reference `ica.certs` in the call to `ca.CreateCSCFromDER`?

Reply:
Fixed in another PR!